rustdoc: word wrap CamelCase in the item list table and sidebar

notriddle opened the pull request:
This is an alternative to https://github.com/rust-lang/rust/pull/126209. That is, it fixes the issue that affects the very long type names in https://docs.rs/async-stripe/0.31.0/stripe/index.html#structs.

This is, necessarily, a pile of nasty heuristics. We need to balance a few issues:

- Sometimes, there's no real word break. For example, `BTreeMap` should be `BTree<wbr>Map`, not `B<wbr>Tree<wbr>Map`.

- Sometimes, there's a legit word break, but the name is tiny and the HTML overhead isn't worth it. For example, if we're typesetting `TyCtx`, writing `Ty<wbr>Ctx` would have an HTML overhead of 50%. Line breaking inside it makes no sense.

# Screenshots

| Before | After |
| ------ | ----- |
| ![image](https://github.com/rust-lang/rust/assets/1593513/d51201fd-46c0-4f48-aee6-a477eadba288) | ![image](https://github.com/rust-lang/rust/assets/1593513/d8e77582-adcf-4966-bbfd-19dfdad7336a)

notriddle committed:
rustdoc: word wrap CamelCase in the item list table

This is an alternative to ee6459d6521cf6a4c2e08b6e13ce3c6ce5d55ed0.
That is, it fixes the issue that affects the very long type names
in https://docs.rs/async-stripe/0.31.0/stripe/index.html#structs.

This is, necessarily, a pile of nasty heuristics.
We need to balance a few issues:

- Sometimes, there's no real word break.
  For example, `BTreeMap` should be `BTree<wbr>Map`,
  not `B<wbr>Tree<wbr>Map`.

- Sometimes, there's a legit word break,
  but the name is tiny and the HTML overhead isn't worth it.
  For example, if we're typesetting `TyCtx`,
  writing `Ty<wbr>Ctx` would have an HTML overhead of 50%.
  Line breaking inside it makes no sense.

notriddle committed:
Fix tidy call in runtest with custom HTML element

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9457008564/job/26049998027) [(plain)](https://github.com/rust-lang/rust/commit/06ce2e08961297d9d1ffa553d949663612b6f747/checks/26049998027/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (120/124)
.FFF       (124/124)


/checkout/tests/rustdoc-gui/label-next-to-symbol.goml label-next-to-symbol... FAILED
[ERROR] `tests/rustdoc-gui/label-next-to-symbol.goml` line 22: XPath "//*[@class='item-name']//a[text()='replaced_function']" not found: for command `compare-elements-position-near: (
    "//*[@class='item-name']//a[text()='replaced_function']",
    ".item-name .stab.deprecated",
    {"y": 2},
)`
[ERROR] `tests/rustdoc-gui/label-next-to-symbol.goml` line 34: XPath "//*[@class='item-name']//a[text()='replaced_function']/.." not found: for command `compare-elements-position: (
    "//*[@class='item-name']//a[text()='replaced_function']/..",
    "//*[@class='desc docblock-short'][text()='a thing with a label']",
    ["y"],
)`
[ERROR] `tests/rustdoc-gui/label-next-to-symbol.goml` line 44: XPath "//*[@class='item-name']//a[text()='replaced_function']" not found: for command `compare-elements-position-near: (
    "//*[@class='item-name']//a[text()='replaced_function']",
    ".item-name .stab.deprecated",
    {"y": 2},
)`
[ERROR] `tests/rustdoc-gui/label-next-to-symbol.goml` line 56: XPath "//*[@class='item-name']//a[text()='replaced_function']/.." not found: for command `compare-elements-position-false: (
    "//*[@class='item-name']//a[text()='replaced_function']/..",
    "//*[@class='desc docblock-short'][text()='a thing with a label']",
    ["y"],

/checkout/tests/rustdoc-gui/notable-trait.goml notable-trait... FAILED
/checkout/tests/rustdoc-gui/notable-trait.goml notable-trait... FAILED
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 11
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 64: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["y"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 17
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 64: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position-false: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["x"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 23
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 64: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            {"x": |x|},
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 72: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position-false: (
    "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
    "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
    ["y", "x"],
)`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 11
    from `tests/rustdoc-gui/notable-trait.goml` line 80: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["y"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 17
    from `tests/rustdoc-gui/notable-trait.goml` line 80: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position-false: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["x"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 23
    from `tests/rustdoc-gui/notable-trait.goml` line 80: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            {"x": |x|},
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 27
    from `tests/rustdoc-gui/notable-trait.goml` line 80: The following errors happened: [different X values: 321.8125 (or 322) != 523]: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            {"x": |i_x|},
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 11
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["y"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 17
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `compare-elements-position-false: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["x"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 23
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: XPath "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']" not found: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[text()='NotableStructWithLongName']",
            {"x": |x|},
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 27
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: The following errors happened: [different X values: 168.625 (or 169) != 293]: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            {"x": |i_x|},

/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 15: Error: Evaluation failed: The following errors happened (for selector `pre.item-decl`): [expected `1324` for property `"scrollWidth"`, found `840`]: for command `assert-property: ("pre.item-decl", {"scrollWidth": "1324"})`
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 19: Error: Evaluation failed: The following errors happened (for selector `.item-table .struct`): [expected `684` for property `"offsetWidth"`, found `165`]: for command `assert-property: (".item-table .struct", {"offsetWidth": "684"})`
Error: ()
Build completed unsuccessfully in 0:03:41
  local time: Tue Jun 11 00:15:35 UTC 2024
  network time: Tue, 11 Jun 2024 00:15:35 GMT
```

</details>

GuillaumeGomez commented:
I'm not a big fan to this approach. Limiting the width of the item name seems better don't you think?

BradMarr commented:
Wouldn't the <wbr> tag only apply if the width is already limited? Wouldn't this be more of an application layer on top of #126209 instead of an alternative?

notriddle committed:
rustdoc: use `<wbr>`-tolerant function to check text contents

notriddle committed:
rustdoc: avoid redundant HTML when there's already line breaks

notriddle committed:
rustdoc: use `<wbr>` in sidebar headers

This also improves sidebar layout, so instead of

    BTreeM
    ap

you get this

    BTree
    Map

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9823866943/job/27122219324) [(plain)](https://github.com/rust-lang/rust/commit/a32cb15f6700827c8c8298d3503c2cc84b7eb623/checks/27122219324/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (120/124)
.FFF       (124/124)


/checkout/tests/rustdoc-gui/notable-trait.goml notable-trait... FAILED
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 11
    from `tests/rustdoc-gui/notable-trait.goml` line 80: different Y values: 302.5 != 326.5: for command `compare-elements-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[normalize-space()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["y"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 27
    from `tests/rustdoc-gui/notable-trait.goml` line 80: The following errors happened: [different X values: 321.8125 (or 322) != 523]: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            {"x": |i_x|},
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 11
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: different Y values: 360.5 != 384.5: for command `compare-elements-position: (
            "//*[@id='method.create_an_iterator_from_read']//a[normalize-space()='NotableStructWithLongName']",
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            ["y"],
        )`
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 27
    from `tests/rustdoc-gui/notable-trait.goml` line 38
    from `tests/rustdoc-gui/notable-trait.goml` line 87: The following errors happened: [different X values: 168.625 (or 169) != 293]: for command `assert-position: (
            "//*[@id='method.create_an_iterator_from_read']//*[@class='tooltip']",
            {"x": |i_x|},

/checkout/tests/rustdoc-gui/sidebar.goml sidebar... FAILED
/checkout/tests/rustdoc-gui/sidebar.goml sidebar... FAILED
[ERROR] `tests/rustdoc-gui/sidebar.goml` line 149: The following errors happened (for selector `.sidebar-elems section .block li > a`): [expected `29` for property `"offsetHeight"`, found `71`]: for command `assert-property: (".sidebar-elems section .block li > a", {"offsetHeight": 29})`
/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 15: The following errors happened (for selector `pre.item-decl`): [expected `1324` for property `"scrollWidth"`, found `840`]: for command `assert-property: ("pre.item-decl", {"scrollWidth": "1324"})`
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 19: The following errors happened (for selector `.item-table .struct`): [expected `684` for property `"offsetWidth"`, found `165`]: for command `assert-property: (".item-table .struct", {"offsetWidth": "684"})`
Error: ()
Build completed unsuccessfully in 0:03:17
  local time: Sun Jul  7 03:02:29 UTC 2024
  network time: Sun, 07 Jul 2024 03:02:29 GMT
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9824504959/job/27123567201) [(plain)](https://github.com/rust-lang/rust/commit/3cabaa2760c1a80347239300d2a38b08f8e50585/checks/27123567201/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (120/124)
..FF       (124/124)


/checkout/tests/rustdoc-gui/sidebar.goml sidebar... FAILED
[ERROR] `tests/rustdoc-gui/sidebar.goml` line 149: The following errors happened (for selector `.sidebar-elems section .block li > a`): [expected `29` for property `"offsetHeight"`, found `71`]: for command `assert-property: (".sidebar-elems section .block li > a", {"offsetHeight": 29})`
/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
/checkout/tests/rustdoc-gui/type-declation-overflow.goml type-declation-overflow... FAILED
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 15: The following errors happened (for selector `pre.item-decl`): [expected `1324` for property `"scrollWidth"`, found `840`]: for command `assert-property: ("pre.item-decl", {"scrollWidth": "1324"})`
[ERROR] `tests/rustdoc-gui/type-declation-overflow.goml` line 19: The following errors happened (for selector `.item-table .struct`): [expected `684` for property `"offsetWidth"`, found `165`]: for command `assert-property: (".item-table .struct", {"offsetWidth": "684"})`
Error: ()
Build completed unsuccessfully in 0:03:03
  local time: Sun Jul  7 04:55:09 UTC 2024
  network time: Sun, 07 Jul 2024 04:55:09 GMT
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #127936) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle committed:
rustdoc: properly handle path wrapping

GuillaumeGomez commented on src/librustdoc/html/escape/tests.rs:
```diff
@@ -0,0 +1,67 @@
+    assert_eq!(&E("first_second").to_string(), "first<wbr>_second");
+    assert_eq!(&E("first:second").to_string(), "first:<wbr>second");
+    assert_eq!(&E("first::second").to_string(), "first::<wbr>second");
+    assert_eq!(&E("MY_CONSTANT").to_string(), "MY<wbr>_CONSTANT");
```
Funnily enough, I would have put the break after the `_` and not before. ^^'

GuillaumeGomez replied:
Awesome, thanks!

notriddle committed:
rustdoc: move the wbr after the underscore, instead of before

GuillaumeGomez commented on tests/rustdoc-gui/duplicate-macro-reexport.goml:
```diff
@@ -4,11 +4,11 @@ go-to: "file://" + |DOC_PATH| + "/test_docs/macro.a.html"
 // Check there is only one macro named "a" listed in the sidebar.
 assert-count: (
-    "//*[@class='sidebar-elems']//*[@class='block macro']//li/a[text()='a']",
+    "//*[@class='sidebar-elems']//*[@class='block macro']//li/a[normalize-space()='a']",
```
In a few cases like this one, are there `<wbr>` tags? Seems surprising, no?

GuillaumeGomez replied:
Ah I see. Please revert these cases then. :)

camelid reviewed (commented):
Looks good on a first pass. Will do another review when I get a chance.

GuillaumeGomez commented on tests/rustdoc/extremely_long_typename.extremely_long_typename.html:
```diff
@@ -0,0 +1 @@
+<li><div class="item-name"><a class="struct" href="struct.CreateSubscriptionPaymentSettingsPaymentMethodOptionsCustomerBalanceBankTransferEuBankTransfer.html" title="struct extremely_long_typename::CreateSubscriptionPaymentSettingsPaymentMethodOptionsCustomerBalanceBankTransferEuBankTransfer">Create<wbr />Subscription<wbr />Payment<wbr />Settings<wbr />Payment<wbr />Method<wbr />Options<wbr />Cus ...
```
I'm surprised that it generates `<wbr />` whereas we generate `<wbr>` in rustdoc. Dark magic.

GuillaumeGomez replied:
Don't worry, I'm not asking to change it or anything. I was just amused to see this change and wondered why this tag was updated this way. But thanks for confirming my suspicions. ;)

GuillaumeGomez commented on tests/rustdoc-gui/type-declation-overflow.goml:
```diff
@@ -70,11 +70,11 @@ go-to: "file://" + |DOC_PATH| + "/lib2/scroll_traits/trait.Iterator.html"
 click: ".item-decl .type-contents-toggle"
 assert-property: ("pre.item-decl", {"scrollLeft": 0})
-scroll-to: "//*[@class='rust item-decl']//details/a[text()='String']"
+scroll-to: "//*[@class='rust item-decl']//details/a[normalize-space()='String']"
```
Is there supposed to have `wbr` in this case too?

GuillaumeGomez commented on tests/rustdoc-gui/source-anchor-scroll.goml:
```diff
@@ -7,11 +7,11 @@ set-window-size: (600, 800)
 assert-property: ("html", {"scrollTop": "0"})
 
-click: '//a[text() = "barbar" and @href="#5-7"]'
+click: '//a[normalize-space() = "barbar" and @href="#5-7"]'
```
Same question here.

GuillaumeGomez commented on tests/rustdoc-gui/sidebar-source-code.goml:
```diff
@@ -65,13 +65,13 @@ go-to: "file://" + |DOC_PATH| + "/src/lib2/another_folder/sub_mod/mod.rs.html"
-assert: "//*[@class='dir-entry' and @open]/*[text()='lib2']"
-assert: "//*[@class='dir-entry' and @open]/*[text()='another_folder']"
-assert: "//*[@class='dir-entry' and @open]/*[text()='sub_mod']"
+assert: "//*[@class='dir-entry' and @open]/*[normalize-space()='lib2']"
```
Same question.

GuillaumeGomez commented on tests/rustdoc-gui/help-page.goml:
```diff
@@ -66,5 +66,5 @@ go-to: "file://" + |DOC_PATH| + "/test_docs/index.html"
 assert-false: "#help"
 click: "#help-button > a"
-click: "//*[@id='help']//a[text()='the rustdoc book']"
+click: "//*[@id='help']//a[normalize-space()='the rustdoc book']"
```
Is it really worth to add `wbr` for this sentence? Even without it, it works fine (can't get any smaller, after that we switch to mobile and the popover doesn't exist there):

![image](https://github.com/user-attachments/assets/a671093c-0973-49be-a421-ccf1be072afe)

GuillaumeGomez commented on src/librustdoc/html/templates/sidebar.html:
```diff
@@ -15,7 +15,9 @@ <h2 class="location"> {# #}
                     {% if !block.heading.name.is_empty() %}
-                        <h3><a href="#{{block.heading.href|safe}}">{{block.heading.name}}</a></h3>
+                        <h3><a href="#{{block.heading.href|safe}}"> {# #}
+                            {{block.heading.name|wrapped|safe}} {# #}
```
I don't think it's useful here considering all the longest headings already have whitespace characters, making them split nicely.

GuillaumeGomez replied:
Hum I see. Then nevermind.

GuillaumeGomez commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit e277d8eb0b21d56aedeabf671c97cc8e85a14fcd has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez e277d8eb0b21d56aedeabf671c97cc8e85a14fcd -->
<!-- homu: {"type":"Approved","sha":"e277d8eb0b21d56aedeabf671c97cc8e85a14fcd","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- 

needs refmt

bors commented:
:pushpin: Commit 25d41526e5597058067be35f126d1029ac078129 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 25d41526e5597058067be35f126d1029ac078129 -->
<!-- homu: {"type":"Approved","sha":"25d41526e5597058067be35f126d1029ac078129","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10147547233/job/28058197852) [(plain)](https://github.com/rust-lang/rust/commit/25d41526e5597058067be35f126d1029ac078129/checks/28058197852/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.907 Building wheels for collected packages: reuse
#16 2.908   Building wheel for reuse (pyproject.toml): started
#16 3.158   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.159   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.159   Stored in directory: /tmp/pip-ephem-wheel-cache-z4lr3m4a/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.162 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.557 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.557 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0432]: unresolved import `askama`
  --> src/librustdoc/html/render/sidebar.rs:83:9
   |
83 |     use askama::{Html, MarkupDisplay};
   |         ^^^^^^ use of undeclared crate or module `askama`
error[E0433]: failed to resolve: use of undeclared crate or module `askama`
  --> src/librustdoc/html/render/sidebar.rs:87:39
   |
   |
87 |     pub(crate) fn wrapped<T>(v: T) -> askama::Result<MarkupDisplay<Html, impl Display>>
   |                                       ^^^^^^ use of undeclared crate or module `askama`
Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `rustdoc` (lib) due to 2 previous errors
Build completed unsuccessfully in 0:02:18
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10147980349/job/28059667065) [(plain)](https://github.com/rust-lang/rust/commit/85b1ddb10a8e1459223ab6145b3c71b0e0d8e125/checks/28059667065/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.761 Building wheels for collected packages: reuse
#16 2.762   Building wheel for reuse (pyproject.toml): started
#16 3.009   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.010   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.010   Stored in directory: /tmp/pip-ephem-wheel-cache-8qqbczdx/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.013 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.406 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.407 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
    Checking arrayvec v0.7.4
   Compiling rinja_derive v0.2.0
    Checking rinja v0.2.0
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0432]: unresolved imports `rinja::Html`, `rinja::MarkupDisplay`
  --> src/librustdoc/html/render/sidebar.rs:83:17
   |
83 |     use rinja::{Html, MarkupDisplay};
   |                 ^^^^  ^^^^^^^^^^^^^ no `MarkupDisplay` in the root
   |                 |
   |                 no `Html` in the root
   = help: consider importing one of these items instead:
           crate::config::OutputFormat::Html
           pulldown_cmark::Event::Html
           pulldown_cmark_old::Event::Html
```

</details>

bors commented:
:pushpin: Commit ac303df4e21eabfbf692f2d1959f7403f4887a31 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez ac303df4e21eabfbf692f2d1959f7403f4887a31 -->
<!-- homu: {"type":"Approved","sha":"ac303df4e21eabfbf692f2d1959f7403f4887a31","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
